library/core/src/iter/traits/iterator.rs

```diff
@@ -1613,6 +1614,100 @@ pub trait Iterator {
+    /// [`PartialEq`] trait implementation.
+    ///
+    /// If the iterator is sorted, this removes all duplicates.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(iter_dedup)]
+    ///
+    /// let vec = vec![1, 2, 2, 3, 2];
+    ///
+    /// let mut iter = vec.into_iter().dedup();
+    ///
+    /// assert_eq!(iter.next(), Some(1));
+    /// assert_eq!(iter.next(), Some(2));
+    /// assert_eq!(iter.next(), Some(3));
+    /// assert_eq!(iter.next(), Some(2));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    #[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+    #[inline]
+    fn dedup(self) -> Dedup<Self, Self::Item>
+    where
+        Self: Sized,
+        Self::Item: PartialEq,
+    {
+        Dedup::new(self)
+    }
+
+    /// Removes all but the first of consecutive elements in the iterator satisfying a given equality
+    /// relation.
+    ///
+    /// The `same_bucket` function is passed a references to two elements from the iterator and
+    /// must determine if the elements compare equal.
+    ///
+    /// If the iterator is sorted, this removes all duplicates.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(iter_dedup)]
+    ///
+    /// let vec = vec!["foo", "bar", "Bar", "baz", "bar"];
+    ///
+    /// let mut iter = vec.into_iter().dedup_by(|a, b| a.eq_ignore_ascii_case(b));
+    ///
+    /// assert_eq!(iter.next(), Some("foo"));
+    /// assert_eq!(iter.next(), Some("bar"));
+    /// assert_eq!(iter.next(), Some("baz"));
+    /// assert_eq!(iter.next(), Some("bar"));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    #[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+    #[inline]
+    fn dedup_by<F>(self, same_bucket: F) -> DedupBy<Self, F, Self::Item>
+    where
+        Self: Sized,
+        F: FnMut(&Self::Item, &Self::Item) -> bool,
+    {
+        DedupBy::new(self, same_bucket)
+    }
+
+    /// Removes all but the first of consecutive elements in the iterator that
+    /// resolve to the same key.
+    ///
+    /// If the iterator is sorted, this removes all duplicates.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(iter_dedup)]
+    ///
+    /// let vec = vec![10, 20, 21, 30, 20];
+    ///
+    /// let mut iter = vec.into_iter().dedup_by_key(|&i| i / 10);
+    ///
+    /// assert_eq!(iter.next(), Some(10));
+    /// assert_eq!(iter.next(), Some(20));
+    /// assert_eq!(iter.next(), Some(30));
+    /// assert_eq!(iter.next(), Some(20));
+    /// assert_eq!(iter.next(), None);
+    /// ```
+    #[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+    #[inline]
+    fn dedup_by_key<F, K>(self, key: F) -> DedupByKey<Self, F, Self::Item>
+    where
+        Self: Sized,
+        F: FnMut(&Self::Item) -> K,
+        K: PartialEq,
+    {
+        DedupByKey::new(self, key)
```

Comment:
```suggestion
        self.dedup_by(|a, b| key(a) == key(b))
```

https://doc.rust-lang.org/stable/src/alloc/vec/mod.rs.html#1441

```rust
        self.dedup_by(|a, b| key(a) == key(b))
```

Reply:
After that change, what is the signature of the function `dedup_by_key`?

Reply:
I think this would work unless I'm forgetting something:

```rust
#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
#[inline]
fn dedup_by_key<F, K, SB>(self, key: F) -> DedupBy<Self, SB, Self::Item>
where
    Self: Sized,
    F: FnMut(&Self::Item) -> K,
    K: PartialEq,
    // "same_bucket"
    SB: FnMut(&Self::Item, &Self::Item) -> bool,
{
    self.dedup_by(|a, b| key(a) == key(b))
}
```

Reply:
You can use the C++ way to give your closure type a name:
```rust
struct EqByKey<F> {
  f: F
}
impl<I, K: PartialEq, F: FnMut(&I) -> K> FnOnce<(&I, &I)> for EqByKey<F> {
  type Output = bool;
  extern "rust-call" fn call_once(mut self, (a, b): (&I, &I)) -> bool {
    (self.f)(a) == (self.f)(b)
  }
}
impl<I, K: PartialEq, F: FnMut(&I) -> K> FnMut<(&I, &I)> for EqByKey<F> {
  extern "rust-call" fn call_mut(&mut self, (a, b): (&I, &I)) -> bool {
    (self.f)(a) == (self.f)(b)
  }
}
```